client/rpc/server/server.go

```diff
@@ -0,0 +1,975 @@
+	return nil
+}
+
+// limitConnections responds with a 503 service unavailable and returns true if
+// adding another client would exceed the maximum allow RPC clients.
+//
+// This function is safe for concurrent access.
+func (s *server) limitConnections(w http.ResponseWriter, remoteAddr string) bool {
+	if int(atomic.LoadInt32(&s.numClients)+1) > s.cfg.RPCMaxClients {
+		log.Infof("Max RPC clients exceeded [%d] - "+
+			"disconnecting client %s", s.cfg.RPCMaxClients,
+			remoteAddr)
+		http.Error(w, "503 Too busy.  Try again later.",
+			http.StatusServiceUnavailable)
+		return true
+	}
+	return false
+}
+
+// incrementClients adds one to the number of connected RPC clients.  Note this
+// only applies to standard clients.  Websocket clients have their own limits
+// and are tracked separately.
+//
+// This function is safe for concurrent access.
+func (s *server) incrementClients() {
+	atomic.AddInt32(&s.numClients, 1)
+}
+
+// decrementClients subtracts one from the number of connected RPC clients.
+// Note this only applies to standard clients.  Websocket clients have their
+// own limits and are tracked separately.
+//
+// This function is safe for concurrent access.
+func (s *server) decrementClients() {
+	atomic.AddInt32(&s.numClients, -1)
+}
+
+// checkAuth checks the HTTP Basic authentication supplied by a wallet or RPC
+// client in the HTTP request r.  If the supplied authentication does not match
+// the username and password expected, a non-nil error is returned.
+//
+// This check is time-constant.
+//
+// The first bool return value signifies auth success (true if successful) and
+// the second bool return value specifies whether the user can change the state
+// of the server (true) or whether the user is limited (false). The second is
+// always false if the first is.
+func (s *server) checkAuth(r *http.Request, require bool) (bool, bool, error) {
+	authhdr := r.Header["Authorization"]
+	if len(authhdr) <= 0 {
+		if require {
+			log.Warnf("RPC authentication failure from %s",
+				r.RemoteAddr)
+			return false, false, errors.New("auth failure")
+		}
+
+		return false, false, nil
+	}
+
+	authsha := sha256.Sum256([]byte(authhdr[0]))
+
+	// Check for limited auth first as in environments with limited users,
+	// those are probably expected to have a higher volume of calls
+	limitcmp := subtle.ConstantTimeCompare(authsha[:], s.limitauthsha[:])
+	if limitcmp == 1 {
+		return true, false, nil
+	}
+
+	// Check for admin-level auth
+	cmp := subtle.ConstantTimeCompare(authsha[:], s.authsha[:])
+	if cmp == 1 {
+		return true, true, nil
+	}
+
+	// Request's auth doesn't match either user
+	log.Warnf("RPC authentication failure from %s", r.RemoteAddr)
+	return false, false, errors.New("auth failure")
+}
+
+// parsedRPCCmd represents a JSON-RPC request object that has been parsed into
+// a known concrete command along with any error that might have happened while
+// parsing it.
+type parsedRPCCmd struct {
```

Comment:
Won't need this after switch to `msgjson`. Also won't need `parseCmd` or `createMarshalledReply`.